runsc/cgroup/cgroup_v2.go

```diff
@@ -0,0 +1,366 @@
+package cgroup
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	cgroupfs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
+	libcontainerconfigs "github.com/opencontainers/runc/libcontainer/configs"
+	libcontainerutils "github.com/opencontainers/runc/libcontainer/utils"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"gvisor.dev/gvisor/pkg/cleanup"
+	"gvisor.dev/gvisor/pkg/log"
+)
+
+type cgroupV2Manager struct {
+	manager libcontainercgroups.Manager
+}
+
+func NewCgroupV2Manager(name string) (*cgroupV2Manager, error) {
+	config := newCgroupConfig(name)
+	manager, err := cgroupfs2.NewManager(config, "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("got manager in NEW %v %s", manager, name)
+	return &cgroupV2Manager{
+		manager: manager,
+	}, nil
+}
+
+// Install creates and configures cgroups according to 'res'. If cgroup path
+// already exists, it means that the caller has already provided a
+// pre-configured cgroups, and 'res' is ignored.
+func (c *cgroupV2Manager) Install(name string, res *specs.LinuxResources) (owned bool, err error) {
+	log.Debugf("Creating cgroup %q", name)
+
+	manager, err := c.getManager(name)
+	if err != nil {
+		return false, err
+	}
+
+	// The Cleanup object cleans up partially created cgroups when an error occurs.
+	// Errors occuring during cleanup itself are ignored.
+	clean := cleanup.Make(func() { _ = c.Uninstall(name, owned) })
+	defer clean.Clean()
+
+	// in unified mode, we only need to look at one path
+	path := buildCgroupUnifiedPath(name)
+	if _, err := os.Stat(path); err == nil {
+		// If cgroup has already been created; it has been setup by caller. Don't
+		// make any changes to configuration, just join when sandbox/gofer starts.
+		log.Debugf("Using pre-created cgroup %q", path)
+	} else {
+		owned = true
+		// Apply(-1) is a hack to create the cgroup directories for each resource
+		// subsystem. The function [cgroups.Manager.apply()] applies cgroup
+		// configuration to the process with the specified pid.
+		// It creates cgroup files for each subsystems and writes the pid
+		// in the tasks file. We use the function to create all the required
+		// cgroup files but not attach any "real" pid to the cgroup.
+		log.Debugf("applied cgroup %q %v %s", path, manager, name)
+		if err := manager.Apply(-1); err != nil {
```

Comment:
...or you can use `manager.CreateCgroupPath()` (not if you want to share the same code with systemd v2 cgroup driver though)